spec/p2p/p2p.qnt

```diff
@@ -0,0 +1,140 @@
+module p2p {
+
+    pure val peerStates = Set("dialing", "inbound", "reconnecting", "outbound", "unconditional")
+    // a work around as constants are not working in the REPL
+
+    pure val connectionStates = Set("init", "open", "closed")
+
+    type Address = {ID:str, ipAddress:str}
+ //   type State = {peerState:str, connectionState: str}
+
+    //type Peer = {address:Address, state:State}
+    //var Peers : Set[Peer]
+
+    // p2p variables 
+    var peerState: Address -> str
+    var connectionState: Address -> str
+    var IsUnconditional: Set[Address]
+
+    // PEX variable
+    var AddressBook: Set[Address]
+
+    // Environment. These are the nodes that may dial us
+    var Environment: Set[Address]
+ 
+
+    pure val MaxNumOutboundPeers = 5
+    pure val MaxNumInboundPeers = 10
+
+    def Peers = peerState.keys()
+
+    def peersOutbound = size(Peers.filter(a => peerState.get(a) == "outbound"))
+    def peersInbound = size(Peers.filter(a => peerState.get(a) == "inbound"))
+    def peersDialing = size(Peers.filter(a => peerState.get(a) == "dialing"))
+
+    def require(pred:bool):bool = pred
+    //pure def putset (m,s)
+
+    // triggered by "the network"
+    action loseConnection(a:Address):bool = all{
+        require(connectionState.get(a) != "closed"),
+        connectionState' = connectionState.set(a, "closed"),
+        all { //unchanged
+            AddressBook' = AddressBook,
+            peerState' = peerState,
+            Environment' = Environment,
+            IsUnconditional' = IsUnconditional,
+        }
+    }
+
+    action transportAccept(a: Address):bool = all {
+        or {
+            peersInbound < MaxNumInboundPeers,
+            IsUnconditional.contains(a)
+        },
+        peerState' = peerState.put(a, "inbound"),  //shall we check if it is already outbound and open?
```

Comment:
Yes, we should. But it can be put on the conditions, right? Peers is not connected.